### ISSUE TITLE
Block fonts fingerprinting

### ISSUE THREAD

**Issue by davidhedlund** [kkapsner/CanvasBlocker#200]
## Description
Custom fonts can be used for fingerprinting.

Solution: Disable custom fonts from loading.

**Comment by davidhedlund** [comment]
https://github.com/kkapsner/CanvasBlocker/issues/162

**Comment by kkapsner** [comment]
I'm not quite sure how I can do that in a WebExtension.
Since this a different attack path I will have to wrap my head around it first. Let's see if I can find a way to protect it.

**Comment by davidhedlund** [comment]
Thank you for your interest in my suggestion. I with you all my luck.

**Comment by spodermenpls** [comment]
I use https://github.com/da2x/fluxfonts to blur the fonts count systemwide.

**Comment by NewGen66** [comment]
Yes on firefox you can limit fonts to single font , it helps

**Comment by spodermenpls** [comment]
@NewGen66 is right, Firefox already provides an option to activate only a selected list of fonts*, so there's not really a need to add this feature to CanvasBlocker as well.

* https://www.soeren-hentzschel.at/firefox/firefox-52-neue-privatsphaere-einstellung-gegen-font-fingerprinting (German)

**Comment by kkapsner** [comment]
I think there isn't even a good way to add this to CB. There is

**Comment by Thorin-Oakenpants** [comment]
First (and I'm not talking about downloadable fonts etc - use uBO, uMatrix etc to block third party eg googlefonts, and @facefont etc), Firefox allows you to:
```
/* disable websites choosing fonts (0=block, 1=allow)
 * [SETTING] General>Language and Appearance>Advanced>Allow pages to choose...
 * [SETTING-ESR52] Content>Font & Colors>Advanced>Allow pages to choose... ***/
user_pref("browser.display.use_document_fonts", 0);
```

This means you fall back on your default set, which can vary depending on whether you use Western, Cyrillic, Japanese/Chinese, Hieroglyphic based languages (not sure of the exact names for these "types")

So for me, being English speaking and all, I use (for my three fonts under options)
```
   // user_pref("font.name.serif.x-unicode", "Georgia");
   // user_pref("font.name.serif.x-western", "Georgia"); // default: Times New Roman
   // user_pref("font.name.sans-serif.x-unicode", "Arial");
   // user_pref("font.name.sans-serif.x-western", "Arial"); // default: Arial
   // user_pref("font.name.monospace.x-unicode", "Lucida Console");
   // user_pref("font.name.monospace.x-western", "Lucida Console"); // default: Courier New
```

others will need to use font.name.serif.x-`cyrillic` etc.

Your default font values will differ as well, depending on your Linux distro. This then leads to icon fonts, also known as glyphs, or tofu (also related are "ligatures"). When your default set of fonts can't map the PUA (its like an address for the font) in your default font , you end up with a hexadecimal tablet symbol. So, for some users, usually Linux, they need to set sans-serif (or maybe it was serif) to Deja Vu or something (otherwise extensions can end up with missing icons as well)

So after all that, you can block basically all font info snatching (we'll just assume you do not use flash), and the web shouldn't complain, but it will get a tad "ugly". I'm saying that quotes, because I have been doing this for at least 3 years, and the web looks perfectly fine to me. You get used to it.

--meanwhile....

The Tor Uplift have two parts for tackling font FP'ing. Part one was completed ages ago (FF52+) and uses this
```
/* 1409: limit system font exposure to a whitelist (FF52+) [SETUP] [RESTART]
 * If the whitelist is empty, then whitelisting is considered disabled and all fonts are allowed.
 * [NOTE] Creating your own probably highly-unique whitelist will raise your entropy. If
 * you block sites choosing fonts in 1401, this preference is irrelevant. In future,
 * privacy.resistFingerprinting (see 4500) may cover this, and 1401 can be relaxed.
 * [1] https://bugzilla.mozilla.org/1121643 ***/
   // user_pref("font.system.whitelist", ""); // (hidden pref)
```

Note it requires a restart. It's a hidden pref (does not show in about:config unless you create it and stick in a value) and has a blank default. People using this right now, are just making themselves super unique. This pref will be used in part two, which they wanted to be ready for ESR60. Hopefully they get it done sooner rather than later. The fonts will be pushed via Kinto.
 * [1336208](https://bugzilla.mozilla.org/show_bug.cgi?id=1336208) bundle & whitelist fonts

They obviously haven't worked out exactly what fonts to bundle, or whether they need to differ for various OS'es (FF's privacy.resistFingerprinting does not hide your OS, but TBB might we have to wait for the final TBB-esr60), but I would assume they would go with what TBB has been using

--so...

If you *did* want to do font anti-FP'ing, expect a world of hurt (esp with cryllic and other non-western/latin ones) in so many ways. Just go look at some of the long (and technical) discussions on tortrac or bugzilla etc. And as kkapsner said, that's even if there's a way to do it. I will add that a previous legacy extension called [FireGloves](https://fingerprint.pet-portal.eu/?menu=6) (it wasn't on AMO, I think because it wasn't allowed due to how it did some injections?) which limited fonts. I remember using it, it broke things - unlike using FF's pref to limit to 3 fonts.

tl;dr: just let tor uplift deal with it all, cuz it's not worth wasting your time spending a year learning about fonts and sh^t :) soz for the long ramble

**Comment by Thorin-Oakenpants** [comment]
^^ https://github.com/kboda/firegloves

**Comment by kkapsner** [comment]
Thanks guys for the valuable input which brings me to the conclusion of not implementing that in CB.

**Comment by atomGit** [comment]
@kkapsner - here's a development that might interest you - it's a very simple extension, but given where it's hosted i (and @Thorin-Oakenpants and others) don't really trust it - if you want to rethink incorporating anti-font fingerprinting, maybe this would help - the add-on gens a unique FP on every page load/reload...

[Font Fingerprint Defender](https://addons.mozilla.org/en-US/firefox/addon/font-fingerprint-defender/)

**Comment by kkapsner** [comment]
You are right not to trust this extension... it's very easy to detect and does not really protect against the fingerprinting (e.g. instead of offset... the fingerprinter could just use client...).

As like other extensions regarding fingerprinting they often only protect against the proof of concept that is available in public but do not think about other attack vectors... and with fonts there are just too many. I think it's better to do this upstream.

**Comment by atomGit** [comment]
thanks for your comment - this confirms what 'pants' felt about it :)
